Updates to Cadence guidance for Solidity developers (#2263)
onflow/cadence: docs/solidity-to-cadence.mdx

Newcomers to Cadence should ensure they understand the following major concepts before development.

## Flow account model

The [Flow account model](learn#concepts#accounts-and-keys) in Cadence combines storage for the keys and code
(”smart contracts”) associated with an account with storage for the assets owned by that account. That’s right:
In Cadence, your tokens are stored in your account, and not in a smart contract. Of course, smart contracts still
define these assets and how they behave, but those assets can be securely stored in a user’s account through the
magic of Resources.

There is only one account type in Cadence also with an account address, similar to an Externally-Owned-Account
(EOA) address in Ethereum. However, unlike Ethereum contract-accounts, accounts in Cadence also store contract code.
Accounts realize ownership on Flow in being the container where keys, Resources, and contracts are stored on-chain.

<Img src="https://storage.googleapis.com/flow-resources/documentation-assets/solidity-to-cadence/account-structure.png" />

### PublicAccount and AuthAccount

> Fields, functions, Resources and/or Capabilities in a contract use [access modifiers](language@access-control)

Reviewer: ```suggestion
Fields, functions, Resources and/or Capabilities in a contract use [access modifiers](https://developers.flow.com/cadence/language/access-control)
```

Reviewer: I've seen there are more of this so probably you are planning to update all at the same time

protecting privileged functionality or making it public. `PublicAccount` is the publicly available part of an account
through which public functions or objects can be accessed by any account using the `getAccount` function.

`AuthAccount` is available only to the signer of the transaction. An `AuthAccount` reference provides full access
to that account's storage, public keys and contract code. Capabilities ensure that resources held in an account can be

> safely shared/accessed; access to `AuthAccount` is never given to other accounts.

Reviewer: ```suggestion
safely shared/accessed; access to `AuthAccount` should never be given to other accounts.
```

I mean, you could pass an `&AuthAcct`, but you shouldn't

## Resources

Resources are unique, [linear-types](https://en.wikipedia.org/wiki/Substructural_type_system#Linear_type_systems) which
can never be copied or implicitly discarded, only moved between accounts. If, during development, a function fails to
store a Resource obtained from an account in the function scope, semantic checks will flag an error. The run-time
enforces the same strict rules in terms of allowed operations. Therefore, contract functions which do not properly
handle Resources in scope before exiting will abort, reverting them to the original storage. These features of
Resources make them perfect for representing tokens, both fungible and non-fungible. Ownership is tracked by where
they are stored, and the assets can’t be duplicated or accidentally lost since the language itself enforces correctness.

Flow encourages storing of data and compute on-chain and Resource-types makes this easier than even. Since Resources
are always stored in accounts, any data that exists in Resource instances is seamlessly being managed on-chain
without any explicit handling needed.

## Capability-based access

> Remote access to stored objects can be delegated via [Capabilities](language#capability-based-access-control). This

Reviewer: ```suggestion
Remote access to stored objects can be managed via [Capabilities](language#capability-based-access-control). This
```

Small nitpick, you don't delegate to someone else the access everytime you use capabilities, you could use them also to restrict how you access your own resources

Capability if it wants to give all other accounts access. (For example, it’s common for an account to accept fungible
token deposits from all sources via a public Capability.) Alternatively, an account can grant private Capabilities
to specific accounts in order to provide access to restricted functionality. For example, an NFT project often
controls minting through an “administrator Capability” that grants specific accounts with the power to mint new tokens.

## Contract standards

There are numerous widely-used contract standards established to benefit the ecosystem, for example
[Fungible Token](flow#core-contracts#fungible-token)(FT) and [Non-Fungible Token](flow#core-contracts#non-fungible-token)(NFT)
standards which are conceptually equivalent to Ethereum's ERC-20 and ERC-721 standards. Cadence's object-oriented
design means standards apply through contract sub-types such as Resources, Resource interfaces, or other types
declared in the contract standard. Standards can define and limit behaviour and/or set conditions which
implementations of the standard cannot violate.

> Detailed information about available standards and other core contracts can be found in [Core contracts](flow#core-contracts).

Reviewer: There's nothing about standards in the core contracts, those are the contracts that govern the chain

Author: It's all combined in this page https://developers.flow.com/flow/core-contracts

Author: Fixed it

<Img src="https://storage.googleapis.com/flow-resources/documentation-assets/solidity-to-cadence/access-based-security.png" />
<Img src="https://storage.googleapis.com/flow-resources/documentation-assets/solidity-to-cadence/capability-based-security.png" />

## Access control using Capabilities

Solidity lacks specific types or other primitives to aid with permission management. Developers have to inline
guards to `require` at every function entry-point, thus validating the `msg.sender` of the transaction.

[Capabilities](language#capability-based-access-control) are defined by linking storage paths (namespaces for contract
storage) to protected objects and then making that linked capability available to other accounts. Public and private
scopes defined for storage paths and Capabilities themselves align precisely with
[PublicAccount](language#accounts#publicaccount)/[AuthAccount](language#accounts#authaccount) account scopes.

Any account can get access to an account's public Capabilities. Public capabilities are created using public paths,
i.e. they have the domain `public`. For example, all accounts have a default public capability linked to the

> `FLOWToken.Vault` Resource. This vault is exposed as a public capability, scoped to the `receiver` resource interface,

Reviewer: ```suggestion
`FlowToken.Vault` Resource. This vault is exposed as a public capability, scoped to the `receiver` resource interface,
```

> to allow any account to `borrow()` the Vault to make a `deposit()`. Public capabilities can be obtained from both

Reviewer: ```suggestion
to allow any account to `borrow()` a reference to the Vault to make a `deposit()`. Since only the methods defined under the `FungibleToken.Receiver` interface are exposed, whoever borrows the reference to the vault won't be able to call the `withdraw()` method that allows to take tokens out of that Vault. Public capabilities can be obtained from both
```

Reviewer: Not 100% sure about the wording here, but it seemed a good place to insist on *how* capabilities make secure to access resources

authorized accounts (`AuthAccount`) and public accounts (`PublicAccount`).

Private capabilities are specifically granted to accounts. They are created using private paths, i.e. they have the

> domain `private`. After creation, they can be obtained from authorised accounts (`AuthAccount`) but not public

Reviewer: ```suggestion
domain `private`. After creation, they can be obtained from authorised account objects (`AuthAccount`) but not public
```

Reviewer: I'll make extra clear that you get access to either AuthAccount or PublicAccount object / reference, but both came from the single type of account that exists on flow (this point is really well explained above, but this line could lead to understand that AuthAccount and PublicAccount are different kinds of accounts)

accounts (`PublicAccount`). To share a private Capability with another account, the owning account must `publish` it
to another account which places in the [account inbox](language#accounts#account-inbox). The recipient can later
claim the Capability from the account inbox using then `claim` function.

Capabilities can be `unpublished` and can also be [revoked](design-patterns#capability-revocation) by the creating
account.

To aid automation, events are emitted for `publish`, `claim` and `unpublish` actions completed for a Capability.

Detailed information can be found in [Capabilities](language#capability-based-access-control).

## Hygiene factors for protecting value

While capabilities grant account access to a protected resource, it's still necessary to impose controls on value

> accessed through them. For example, if your use-case requires delegating access to a `FLOWToken.Vault` to

Reviewer: ```suggestion
accessed through them. For example, if your use-case requires delegating access to a `FlowToken.Vault` to
```

`withdraw()` funds, it's important to limit the amount. Tokens implementing FT/NFT standards are the primary type of
value being exchanged by accounts on Flow. The standard provides the primitives needed to implement capability
limiting best-practices.

### Token isolation

All FTs reside in a `Vault` Resource and each different FT will exist as a separate `Vault` in an account. Similarly,
all NFTs implement a `Collection` Resource, in which those NFTs held by an account for that collection are stored.

Whenever access to the `withdraw()` function has to be delegated to another account, the simplest way to limit how
many tokens of a given type can be withdrawn is to create a new `Vault` Resource for that token type and move a smaller
amount of the tokens in the main token `Vault`. A capability is then linked to that `Vault` instance before being
made available to another account.

> A similar pattern can be used for NFTs, where a new `Collection` Resource can be created into which only those NFTs

Reviewer: This could be done, but I haven't really seen any piece of code doing it. I believe it's really well explained just with the FT paragraph

in accounts.

## Guard Check

Solidity uses `revert`, `require` & `assert` to validate inputs. `require` is a product of the address-based nature
of Solidity which Capabilities replace. `revert` is similar to Cadence's `panic` in that a transaction is aborted.
Cadence provides an `assert` operator which mirrors `assert` in Solidity.

## Modifiers

Modifiers are extensively used in Solidity when enforcing pre-checks within a function. This is a powerful language
feature. However, modifiers can also mutate state which introduces risks to program control flow.

Cadence uses `pre` and `post` blocks to validate input values or the function execution outputs. Notably, `pre` and
`post` block prohibit changing of state and may only enforce conditions.

Another difference is that modifiers in Solidity can be re-used within the contract multiple times. Cadence
`pre` & `post` blocks are associated to individual functions only, reducing the likelihood of errors but which
results in a small amount of code duplication.

## Error handling

Solidity offers try/catch block to handle errors, however, there is no equivalent in Cadence. This is because

> Resources and Capabilities codify the complexity of access and security in first-class types and

Reviewer: Missing finish of the sentence. Anyway I'll maybe skip this point since I believe there's a chance that we get some kind of try / catch in cadence in the future (not really sure about this, just a vague memory I have about some conversations)

Author: Good catch, will clean it up. Thx

Cadence implements [saturating math](https://en.wikipedia.org/wiki/Saturation_arithmetic) that avoids
overflow/underflow.

## Optional support

[Optional bindings](language#control-flow) provide in-built conditional handling of nil values. Optionals have two
cases: either there is a value, or there is nothing; they fork program flow based on whether a value is nil or not.

## Iterable Dictionaries

Solidity offers the mapping type, however, it is not iterable. Because of that dApp developers have to run off-chain
services to know all the keys to the mapping. This also pushes builders to create custom datatypes like EnumerableMap
which adds to gas costs.

Cadence offers the [Dictionary](language#control-flow) type, an unordered collection of key-value associations
which is iterable.

## Rich support for type utility functions

Cadence offers numerous native-type utility functions to simplify development. For example, the String type
provides:
- utf8
- concat()
- slice()
- decodeHex()
- encodeHex()
- toLower()
- length

## Argument labelling

Argument labels in Cadence help to disambiguate input values. They makes code more readable and explicit. They

> also eliminate confusion around the order of arguments when working with the same type.

Reviewer: ```suggestion
also eliminate confusion around the order of arguments when working with the same type. They have to be included on the function call. This restriction can be skipped if the label is preceded by `_ ` on its declaration.
```

Eg:
`fun foo(balance: UFix64)`
called as

> `self.foo(balance: 30.0)`

Reviewer: ```suggestion
`self.foo(balance: 30.0)`

`fun foo( _balance: UFix64)`
can be called as
`self.foo(balance: 30.0)`
or as
`self.foo(30.0)`
```

Maybe this is going too much into detail on cadence syntax, however I'll make an extra emphasis on that argument labels are mandatory when calling the function by default

address), resulting in all contracts having to carefully track and evaluate access and authorizations.

<img src="https://storage.googleapis.com/flow-resources/documentation-assets/solidity-to-cadence/ethereum-ownership.png" width="450" align="right"/>

Transactions are based on who authorized them, which is provided as `msg.sender` in the transaction context.
User to contract, or contract to contract interactions must be explicitly coded to ensure the appropriate approvals
have been made before interacting with a contract. The contract based nature of storage means that user ownership
in Ethereum is represented in a mapping, for example from owner to balance, or token ID to owner. Put another
way, ownership is tracked in ledger records similar to a person's bank balance. Crypto wallets help combine
balances from multiple token types into a convenient view for the user.

Cadence introduces new primitives and distinct functionalities, namely Resources and Capabilities, that are designed
around Flow's account model. Resources are first-class language types which are unique, non-copyable, and which cannot

> be discarded. These properties make Resources ideal for representing tokens that are always limited in numbers.

Reviewer 2: ```suggestion
be discarded. These properties make Resources ideal for representing digital assets like currency or tokens that are always limited in numbers.
```

(”smart contracts”) associated with an account with storage for the assets owned by that account. That’s right:
In Cadence, your tokens are stored in your account, and not in a smart contract. Of course, smart contracts still
define these assets and how they behave, but those assets can be securely stored in a user’s account through the
magic of Resources.

<img src="https://storage.googleapis.com/flow-resources/documentation-assets/solidity-to-cadence/account-structure.png"  width="300" align="right" />

There is only one account type in Cadence also with an account address, similar to an Externally-Owned-Account
(EOA) address in Ethereum. However, unlike Ethereum contract-accounts, accounts in Cadence also store contract code.
Accounts realize ownership on Flow in being the container where keys, Resources, and contracts are stored on-chain.

### PublicAccount and AuthAccount

Fields, functions, Resources and/or Capabilities in a contract use [access modifiers](language#access-control)
protecting privileged functionality or making it public. `PublicAccount` is the publicly available part of an account

> through which public functions or objects can be accessed by any account using the `getAccount` function.

Reviewer 2: you mention access modifiers, then `PublicAccount`, which are two different forms of access control. Probably just remove the first sentence and start this section with the section about `PublicAccount

`AuthAccount` is available only to the signer of the transaction. An `AuthAccount` reference provides full access

> to that account's storage, public keys and contract code. Capabilities ensure that resources held in an account can be

Reviewer 2: ```suggestion
to that account's storage, keys configuration, and contract code. Capabilities ensure that resources held in an account can be
```

safely shared/accessed; access to `AuthAccount` should never be given to other accounts.

## Resources

Resources are unique, [linear-types](https://en.wikipedia.org/wiki/Substructural_type_system#Linear_type_systems) which
can never be copied or implicitly discarded, only moved between accounts. If, during development, a function fails to
store a Resource obtained from an account in the function scope, semantic checks will flag an error. The run-time
enforces the same strict rules in terms of allowed operations. Therefore, contract functions which do not properly
handle Resources in scope before exiting will abort, reverting them to the original storage. These features of
Resources make them perfect for representing tokens, both fungible and non-fungible. Ownership is tracked by where
they are stored, and the assets can’t be duplicated or accidentally lost since the language itself enforces correctness.

> Flow encourages storing of data and compute on-chain and Resource-types makes this easier than even. Since Resources

Reviewer 2: ```suggestion
Flow encourages storing of data and compute on-chain and Resource-types makes this easier than ever. Since Resources
```

> are always stored in accounts, any data that exists in Resource instances is seamlessly being managed on-chain

Reviewer 2: ```suggestion
are always stored in accounts, any data that exists in Resource instances is seamlessly managed on-chain
```

declared in the contract standard. Standards can define and limit behaviour and/or set conditions which
implementations of the standard cannot violate.

Detailed information about available standards and other core contracts can be found in [Introduction to Flow](../flow).

### NFT standard and metadata

Solidity must manage NFT metadata off-chain and NFTs typically link to IPFS JSON from on-chain.

The Cadence NFT standard provides in-built support for metadata with specific types called [views](../flow#flow-nft#overview#nft-metadata).
Views can be added to NFTs when minted and will always be available as part of the NFT. While metadata is stored
on-chain, graphics and video content are stored off-chain. Cadence provides [utility views](../flow#flow-nft#overview#list-of-common-views)
for both HTTP and IPFS based media storage which remain linked to your NFT.

Using NFT metadata views is a requirement to get listed in the [Flow NFT Catalog](https://www.flow-nft-catalog.com/).
Projects are encouraged leverage the NFT catalog since wallets and other ecosystem partners can seamlessly integrate
new collections added there with no input from project creators.

> NFT metadata on Flow opens the door to exciting new possibilities that help builder to innovate. Check out this

Reviewer 2: ```suggestion
NFT metadata on Flow opens the door to exciting new possibilities that help builders innovate. Check out this
```

new collections added there with no input from project creators.

NFT metadata on Flow opens the door to exciting new possibilities that help builder to innovate. Check out this
recent [case study](https://flow.com/post/flovatar-nft-flow-blockchain-case-study) where a community partner
leveraged SVG on-chain NFT metadata to make combined 2D + 3D versions of their PFPs, all on-chain inside the NFT
metadata!

# Security and access control

Decentralized application development places significant focus on security and access and can fairly be described
as security engineering. Understanding how Resources, Capabilities and the account model solve this may not
be obvious when viewed from a Solidity perspective.

## msg.sender considered harmful

The first question that every Solidity developer asks when they start programming in Cadence is:

**"How do I get the account who authorized the transaction?"**

In Ethereum this account is referred to as `msg.sender` and informs the program flow in a function depending on who
authorized it. Doing so is key to access and security, and is the basis of identity and ownership on Ethereum.

> Cadence does not have `msg.sender` and there is no transaction-level way for Cadence contract to uniquely identify the

Reviewer 2: ```suggestion
Cadence does not have `msg.sender` and there is no transaction-level way for Cadence code to uniquely identify the
```

calling account. Even if there was a way to access it, Cadence supports [multi-sig](#multi-key-multi-signature-support)

> transactions, meaning that a list of all the signers' accounts would be returned.

Reviewer 2: ```suggestion
transactions, meaning that a list of all the signers' accounts would be returned, making it impossible to identify a single authorizer.
```

The [capability-based security](https://en.wikipedia.org/wiki/Capability-based_security) model frames access in
the opposite direction than the [access-based security](https://en.wikipedia.org/wiki/Access-control_list) model.

## Access control using Capabilities

Solidity lacks specific types or other primitives to aid with permission management. Developers have to inline
guards to `require` at every function entry-point, thus validating the `msg.sender` of the transaction.

[Capabilities](language#capability-based-access-control) are defined by linking storage paths (namespaces for contract
storage) to protected objects and then making that linked capability available to other accounts. Public and private
scopes defined for storage paths and Capabilities themselves align precisely with
[PublicAccount](language#accounts#publicaccount)/[AuthAccount](language#accounts#authaccount) account scopes.

Any account can get access to an account's public Capabilities. Public capabilities are created using public paths,
i.e. they have the domain `public`. For example, all accounts have a default public capability linked to the

> `FlowToken.Vault` Resource. This vault is exposed as a public capability, scoped to the `receiver` resource interface,

Reviewer 2: ```suggestion
`FlowToken.Vault` Resource. This vault is exposed as a public capability, scoped to the `FungibleToken.Receiver` resource interface,
```

Reviewer 2: might be good to have a link to the interface or just paste the `Receiver` interface into here to show what it is
